### PR TITLE
[Mg26x] WIP: Added NLO cards for ZJets/WJets processes

### PR DESCRIPTION
**Pull request by cmantill**
Hi, I'm trying to work on creating NLO gridpacks for ZJets/WJetstoQQ processes with the bias weight functionality and validate them with those without bias weight. Since I don't have any previous experience working with NLO cards can somebody help me review the syntax on the cards before proceeding. Also any input about how to validate them would be appreciated. Thanks!

### CODE REVIEW

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX/ZJetsToQQ_5f_NLO_FXFX_run_card.dat [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+#***********************************************************************
+# Number of events (and their normalization) and the required          *
+# (relative) accuracy on the Xsec.                                     *
+# These values are ignored for fixed order runs                        *
+#***********************************************************************
+  1500 = nevents ! Number of unweighted events requested 
+ 0.001 = req_acc ! Required accuracy (-1=auto determined from nevents)
+    20 = nevt_job! Max number of events per job in event generation. 
+                 !  (-1= no split).
+average = event_norm ! Normalize events to sum or average to the X sect.
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX/ZJetsToQQ_5f_NLO_FXFX_run_card.dat]: change the average to bias

**Comment by menglu21** [comment]
hi,cmantill
the file you need to modify is "SubProcess/cut.f", the subroutine bias_weight_function, change the run_card, modify average = event_norm to bias=event_norm in the mean time. of course you could get some simple results using lhe events, but the better way you should use corresponding gridpack to generate some events, and do some simple check with the inclusive MC sample to validate that the bias weight works

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX/ZJetsToQQ_5f_NLO_FXFX_run_card.dat (hunk 1 above)
Comment (cmantill) [reply bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX/ZJetsToQQ_5f_NLO_FXFX_run_card.dat]: That's already done here https://github.com/cmantill/genproductions/tree/mg26x/bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX_bias 

I uploaded both sets of cards (with and without bias weight) - I'm not sure this has been the inclusive one has been produced before (or if so I could not find the old cards for NLO).

**Comment by cmantill** [comment]
Hi Meng, 
Indeed, I modified the run_card here https://github.com/cmantill/genproductions/blob/mg26x/bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX_bias/ZJetsToQQ_5f_NLO_FXFX_bias_run_card.dat#L32 and
the cuts.f
https://github.com/cmantill/genproductions/blob/mg26x/bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX_bias/ZJetsToQQ_5f_NLO_FXFX_bias_cuts.f#L922-L957

**Comment by menglu21** [comment]
Hi @cmantill , i think the cards are fine. have you firstly done some plots using lhe events?
actually, i am trying to do the validation using some more simple process like p p > l+ l- j[QCD], as there are existing samples.

**Comment by cmantill** [comment]
Hi @menglu21 , I've never done plots using lhe events before, can you point me to a similar script that uses lhe files? In any case, I should create the gridpack before doing that right?

**Comment by cmantill** [comment]
Actually, when trying to produce the gridpacks I get the following error: 
Error detected in sub-command output WJetsToQQ_5f_NLO_FXFX_bias -nojpeg
write debug file MG5_debug
If you need help with this issue please contact us on https://answers.launchpad.net/mg5amcnlo
MadGraph5Error : A compilation Error occurs when trying to compile /afs/cern.ch/work/c/cmantill/public/Gen/up/genproductions/bin/MadGraph5_aMCatNLO/WJetsToQQ_5f_NLO_FXFX_bias/WJets\
ToQQ_5f_NLO_FXFX_bias_gridpack/work/MG5_aMC_v2_6_0/vendor/IREGI/src.

Examples of full logs are here:
http://cmantill.web.cern.ch/cmantill/WJetsToQQ_5f_NLO_FXFX_bias.log and
http://cmantill.web.cern.ch/cmantill/ZJetsToQQ_5f_NLO_FXFX.log

**Comment by menglu21** [comment]
hi, for lhe plot, you can have a look on https://github.com/menglu21/Gen_production/blob/master/draw_ptj.C
i think it's an efficient way to check your cards by launching your cards to generate some lhe events.

**Comment by menglu21** [comment]
hi, i think the problem is that "MadGraph5Error : A compilation Error occurs when trying to compile /afs/cern.ch/work/c/cmantill/public/Gen/up/genproductions/bin/MadGraph5_aMCatNLO/WJetsToQQ_5f_NLO_FXFX_bias/WJetsToQQ_5f_NLO_FXFX_bias_gridpack/work/MG5_aMC_v2_6_0/vendor/IREGI/src.
	The compilation fails with the following output message:
	    (cd /afs/cern.ch/user/c/cmantill/work/public/Gen/up/genproductions/bin/MadGraph5_aMCatNLO/WJetsToQQ_5f_NLO_FXFX_bias/WJetsToQQ_5f_NLO_FXFX_bias_gridpack/work/vendor/IREGI/src;make clean -f makefile_ML5_lib;make -f makefile_ML5_lib)
	    /bin/sh: line 0: cd: /afs/cern.ch/user/c/cmantill/work/public/Gen/up/genproductions/bin/MadGraph5_aMCatNLO/WJetsToQQ_5f_NLO_FXFX_bias/WJetsToQQ_5f_NLO_FXFX_bias_gridpack/work/vendor/IREGI/src: No such file or directory"
it seems that you'd better use public directory to work as there are public directory and they are different.

**Comment by perrozzi** [comment]
any news on this?

**Comment by cmantill** [comment]
Hi @perrozzi, sorry just got back to this, it seems that now my gridpacks are running smoothly, will wait for them to finish (hopefully < 1 day) and post here validation plots/update when they are ready. Thanks, Cristina

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/WJetsToQQ_5f_NLO_FXFX/WJetsToQQ_5f_NLO_FXFX_madspin_card.dat [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#
+#directory for gridpack mode
+set ms_dir ./madspingrid
+
+#initialization parameters
+ set Nevents_for_max_weigth 250 # number of events for the estimate of the max. weight
+# set BW_cut 15                # cut on how far the particle can be off-shell
+ set max_weight_ps_point 400  # number of PS to estimate the maximum for each event
+
+# BW_cut = 40 required by Hbb
```
Comment (perrozzi) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/WJetsToQQ_5f_NLO_FXFX/WJetsToQQ_5f_NLO_FXFX_madspin_card.dat]: please remove commented instructions and more generally please follow these recomendations https://twiki.cern.ch/twiki/bin/view/CMS/Moriond18MC#Checks_to_the_datacards_BEFORE_T

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/WJetsToQQ_5f_NLO_FXFX/WJetsToQQ_5f_NLO_FXFX_run_card.dat [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+#***********************************************************************
+    1   = lpp1    ! beam 1 type (0 = no PDF)
+    1   = lpp2    ! beam 2 type (0 = no PDF)
+ 6500   = ebeam1  ! beam 1 energy in GeV
+ 6500   = ebeam2  ! beam 2 energy in GeV
+#***********************************************************************
+# PDF choice: this automatically fixes also alpha_s(MZ) and its evol.  *
+#***********************************************************************
+ lhapdf    = pdlabel   ! PDF set                                     
+ 306000    = lhaid     ! if pdlabel=lhapdf, this is the lhapdf number
```
Comment (perrozzi) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/WJetsToQQ_5f_NLO_FXFX/WJetsToQQ_5f_NLO_FXFX_run_card.dat]: use the recommended substitutes for the actual lhaid number https://twiki.cern.ch/twiki/bin/view/CMS/Moriond18MC#Checks_to_the_datacards_BEFORE_T

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX/ZJetsToQQ_5f_NLO_FXFX_madspin_card.dat [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+#
+#directory for gridpack mode
+set ms_dir ./madspingrid
+
+#initialization parameters
+ set Nevents_for_max_weigth 250 # number of events for the estimate of the max. weight
+# set BW_cut 15                # cut on how far the particle can be off-shell
+ set max_weight_ps_point 400  # number of PS to estimate the maximum for each event
+
+# BW_cut = 40 required by Hbb
```
Comment (perrozzi) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX/ZJetsToQQ_5f_NLO_FXFX_madspin_card.dat]: please remove commented commands (everywhere)

**Comment by perrozzi** [comment]
any news on this?

**Comment by cmantill** [comment]
I corrected all the comments that were missing. I have also already produced the gridpacks with and without bias weight with the current cards, to start validation.

I have no experience doing LHE plots but I tried picking up the information directly from the LHE file and plotting some quantities, e.g. pT of Z: http://cmantill.web.cern.ch/cmantill/ZJets/Dec12/pTZ_log.png ( blue is biased, and green is without weight).  If you have any other scripts to read LHE files that I could start to use to validate the samples it would be very useful. What kind of validation is expected for the bias samples? Is this still the way to go or do we plan to produce HT or pT bins of the NLO samples instead?

Thanks!

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX/ZJetsToQQ_5f_NLO_FXFX_run_card.dat (hunk 1 above)
Comment (perrozzi) [reply bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX/ZJetsToQQ_5f_NLO_FXFX_run_card.dat]: I think this change is still needed, cfr https://github.com/cms-sw/genproductions/pull/1621/files#diff-2c6138a3a9aaa18acaa48176ae3137dbR32

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/WJetsToQQ_5f_NLO_FXFX/WJetsToQQ_5f_NLO_FXFX_run_card.dat [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+#***********************************************************************
+# Number of events (and their normalization) and the required          *
+# (relative) accuracy on the Xsec.                                     *
+# These values are ignored for fixed order runs                        *
+#***********************************************************************
+  1500 = nevents ! Number of unweighted events requested 
+ 0.001 = req_acc ! Required accuracy (-1=auto determined from nevents)
+    20 = nevt_job! Max number of events per job in event generation. 
+                 !  (-1= no split).
+average = event_norm ! Normalize events to sum or average to the X sect.
```
Comment (perrozzi) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/WJetsToQQ_5f_NLO_FXFX/WJetsToQQ_5f_NLO_FXFX_run_card.dat]: also here https://github.com/cms-sw/genproductions/pull/1621/files#diff-2c6138a3a9aaa18acaa48176ae3137dbR32

Comment (cmantill) [reply bin/MadGraph5_aMCatNLO/cards/production/13TeV/WJetsToQQ_5f_NLO_FXFX/WJetsToQQ_5f_NLO_FXFX_run_card.dat]: These changes were already done in the "bias" cards. So if you see the pull request contains 4 folders: 2 are Zjets and Wjets at NLO but with no bias implementation and the other 2 are with bias i.e. these ones:

https://github.com/cmantill/genproductions/tree/mg26x/bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZJetsToQQ_5f_NLO_FXFX_bias

and 

https://github.com/cmantill/genproductions/tree/mg26x/bin/MadGraph5_aMCatNLO/cards/production/13TeV/WJetsToQQ_5f_NLO_FXFX_bias

**Pull request merged**